compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -419,6 +516,8 @@ impl<'a> Parser<'a> {
             // they are dealt with later in resolve.
             self.parse_pat_ident(BindingAnnotation::NONE, syntax_loc)?
         } else if self.is_start_of_pat_with_path() {
+            let snapshot = self.create_snapshot_for_diagnostic();
```

Comment:
Snapshot creation is relatively costly since it clones the entire parser (that's 264 bytes on 64-bit machines) and this is a *good path* (non-error code path) if I read things correctly.

If feasible could you make snapshot creations conditional on `may_recover()` and `look_ahead(1, …)`? Try to *delay* the snapshot creation as far as possible. If you manage to move the creation into the error path, then `look_ahead` is not strictly necessary.

Since you create snapshots in several places in this PR, could you double-check if each one can be improved? Thanks!